include/zephyr/sys/async.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * @brief Public APIs for Async utilities
+ *
+ * Functions and datatypes useful for safely communicating to a thread or work
+ * queue of an asynchronous functions completion.
+ */
+
+#ifndef ZEPHYR_INCLUDE_ASYNC_H_
+#define ZEPHYR_INCLUDE_ASYNC_H_
+
+#include <zephyr/kernel.h>
+
+/**
+ * @brief Async operation callback type
+ *
+ * Common function pointer type used for asynchronous completion callbacks
+ * and implemented for some commonly used IPC notification mechanisms.
+ *
+ * May also be implemented by an application with the understanding
+ * that the callback is most likely to be done in the ISR context
+ * with the limitations and restrictions that imposes.
+ *
+ * @param result The result of the asynchronous call
+ * @param userdata Generic userdata requested to be tied
+ *        to the callback.
+ */
+typedef void (*async_callback_t)(int result, void *userdata);
+
+#ifdef CONFIG_ASYNC
+
+/**
+ * Semaphore completion object with return code.
+ */
+struct async_sem {
+	/** Semaphore that will be given on completion */
+	struct k_sem *sem;
+
+	/** Result of async call */
+	int result;
+};
+
+/**
+ * Asynchronous semaphore completion callback
+ *
+ * Sets the result and gives the semaphore
+ *
+ * @param result Result to assign to the async_sem
+ * @param userdata Pointer to a struct async_sem
+ */
+void async_semaphore_cb(int result, void *userdata);
```

Comment:
I see the idea behind using `userdata` as the argument name, but it can be a bit confusing - this really isn't "user data" as the user is not free to pass arbitrary data pointers here (which goes against all the existing places we have a `void *user_data` argument in callbacks) - it must be a `struct async_sem *`.

I suggest naming the parameter to reflect the data type to be passed. This goes for all the callbacks.

Reply:
I think each of these helper functions needs to state which context they can be called from (kernel, isr, userspace, ..?). Right now, I think all support being called from isr context, but this may change down the road when more helpers are added.